jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/ReservedThreadExecutor.java

```diff
@@ -179,27 +149,9 @@ public void doStop() throws Exception
 
         super.doStop();
 
-        // Mark this instance as stopped.
-        int size = _count.getAndSetLo(-1);
-
-        // Offer the STOP task to all waiting reserved threads.
-        for (int i = 0; i < size; ++i)
-        {
-            // Yield to wait for any reserved threads that
-            // have incremented the size but not yet polled.
-            Thread.yield();
-            _queue.offer(STOP);
-        }
-
-        // Interrupt any reserved thread missed the offer,
-        // so they do not wait for the whole idle timeout.
-        _threads.stream()
-            .filter(ReservedThread::isReserved)
-            .map(t -> t._thread)
-            .filter(Objects::nonNull)
-            .forEach(Thread::interrupt);
-        _threads.clear();
-        _count.getAndSetHi(0);
+        _threads.removeAll().forEach(ReservedThread::stop);
+        Thread.yield();
+        _threads.removeAll().forEach(ReservedThread::stop);
```

Comment:
This feels yucky.

We used the `AtomicBiInteger` to keep track of whether the RTE was stopped, perhaps we can still do that?

So that we would be able to atomically shut down `ThreadIdPool`, similar to `ExecutorService.shutdownNow()` that also returns a `List`?

Reply:
The main loop of a ReservedThread is checking `isRunning` so it should stop when woken.  Adding another atomic doesn't help... it still suffers from test then wait race. We just need to ensure that all threads are woken.

I was just concerned with threads that were missed by the first removeAll because they were just about to be added to the pool and had already done the `isRunning` check.  The second loop should get those an no others should be able to run the same race because they have to do an `isRunning` check.    So I think this code is good.

Reply:
@gregw either this is thread safe and so there is no need to `yield()` and `removeAll()` again, or it is not thread safe and no matter how many times we yield and removeAll() won't fix the problem.

I can see that the combination of read the volatile `_thread` in `stop()` and interrupting it could be enough, so I like much better just one call to `removeAll()`.

Reply:
@sbordet Thinking about it.... this is not really thread safe.  The problem is that the reserved thread does:
```java
                while (isRunning())
                {
                    int slot = _threads.offer(this);
                    // ...
                    Runnable task = waitForTask();
                    // ...
                 }
```

and `doStop()` does
```java
        super.doStop();

        _threads.removeAll().forEach(ReservedThread::stop);
        Thread.yield();
        _threads.removeAll().forEach(ReservedThread::stop);
```

So if the reservedThread is preempted after checking `isRunning`, then `super.doStop()` is called, then the reserved thread might not have put itself in the pool before the first `removeAll`.  So it will not be interrupted and it will go on to `waitForTask`

The idea is that the `yield` will allow that reserve thread to progress, so that the second `removeAll` will see it.....  But I'm not sure that is true.... it could remain preempted between `isRunning` and `offer` over a yield, so it might not be seen by the second remove all.

I think in 99.99999% of cases, yield will be enough.... but it is not 100%

Perhaps we should do the `isRunning` check after the `offer`, but before the wait.  That way either the reserved thread will be seen by the removeAll, or it will see `!isRunning`. 

I'll try that....